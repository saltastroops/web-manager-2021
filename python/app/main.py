from fastapi import FastAPI
from starlette.requests import Request
from starlette.responses import Response
from starlette.templating import Jinja2Templates

app = FastAPI()

templates = Jinja2Templates(directory="templates")


@app.get("/")
def home(request: Request) -> Response:
    return templates.TemplateResponse("home.html", {"request": request})
