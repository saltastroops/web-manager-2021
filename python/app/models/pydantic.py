import asyncio
from asyncio import Task
from typing import TYPE_CHECKING, cast

if TYPE_CHECKING:  # avoid circular import issues
    from app.util.permission import Permission  # pragma: no cover
    from app.util.role import Role  # pragma: no cover

from pydantic import BaseModel


class AccessToken(BaseModel):
    access_token: str
    token_type: str


class User(BaseModel):
    email: str
    family_name: str
    given_name: str
    username: str

    async def has_role_of(self, role: "Role") -> bool:
        """
        Check whether the user has a role.

        If you need to check whether the user has at least one of a list of roles, you
        should use the method has_any_role_of instead of calling has_role_of multiple
        times.

        This method uses the role's is_assigned_to method for performing the check.
        """
        return await role.is_assigned_to(self)

    async def has_any_role_of(self, *roles: "Role") -> bool:
        """
        Check whether the user has any of a list of roles.

        False is returned if the list of roles is empty.

        If you need to check whether the user has at least one of a list of roles, you
        should use this method rather than calling the has_role_of method for each of
        the roles.
        """

        if len(roles) == 0:
            return False

        # asyncio.wait must be used with tasks
        tasks = [asyncio.create_task(self.has_role_of(role)) for role in roles]

        # loop until the user has a role or all roles have been checked
        while True:
            done, pending = await asyncio.wait(
                tasks, return_when=asyncio.FIRST_COMPLETED
            )

            # check whether the user has any of the roles whose checks have completed
            for done_task in done:
                if done_task.result() is True:
                    # the user has a role, so there is no need to continue checking
                    for pending_task in pending:
                        pending_task.cancel()
                    return True

                # the completed tasks are removed, as otherwise we would loop needlessly
                # over the while loop while while waiting for the next task to complete
                tasks.remove(cast(Task[bool], done_task))

            # all checks have completed, and the user has none of the roles
            if not pending:
                return False

    async def is_permitted_to(self, permission: "Permission") -> bool:
        """
        Check whether this user has a permission.

        This method uses the permission's is_permitted_for method for performing the
        check.
        """
        return await permission.is_permitted_for(self)


class UserInDB(User):
    hashed_password: str
