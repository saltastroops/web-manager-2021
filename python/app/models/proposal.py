from datetime import date, datetime
from typing import List, Optional, Dict, Any

from astropy.coordinates import Angle
from astropy.units import Quantity
from pydantic import BaseModel, validator

from app.models.general import Semester


class TextContent(BaseModel):
    """
    The text content of a proposal.

    Fields
    ------
    title:
        Proposal title.
    abstract:
        Proposal abstract.
    read_me:
        Instructions for the observer.
    nightlog_summary:
        Brief (one-line) summary to include in the nightlog.
    """

    title: str
    abstract: str
    read_me: str
    nightlog_summary: str


class Partner(BaseModel):
    """
    A SALT partner.

    Fields
    ------
    code:
        Partner code, such as IUCAA or RSA.
    name:
        Partner name.
    """

    code: str
    name: str


class Institute(BaseModel):
    """
    An institute, optionally with a department.

    Fields
    ------
    partner:
        SALT Partner to which the institute belongs.
    name:
        Institute name.
    department:
        Department within the institute.
    home_page:
        URL of the institute's (or department's) home page.
    """

    partner: Partner
    name: str
    department: Optional[str]
    home_page: str


class PersonalDetails(BaseModel):
    """
    Personal details.

    Fields
    ------
    given_name:
        The given ("first") name(s).
    family_name:
        The family ("last") name.
    email:
        The preferred email address.
    """

    given_name: str
    family_name: str
    email: str


class Investigator(BaseModel):
    """
    An investigator on a proposal.

    Fields
    ------
    is_pc:
        Whether the investigator is the Principal Contact.
    is_pi:
        Whether the investigator is the Principal Investigator.
    personal_details:
        Personal details of the investigator, which may differ from the ones given in
        the proposal.
    affiliation:
        Affiliation of the investigator (for the proposal).
    """

    is_pc: bool
    is_pi: bool
    personal_details: PersonalDetails
    affiliation: Institute


class TimeAllocations(BaseModel):
    """
    A time allocation by a partner.

    Fields
    ------
    partner:
        SALT partner that allocated the time.
    tac_comment:
        An (optional) comment made by the TAC on the proposal for which this time
        allocation was made.
    priority_0:
        Allocated priority 0 time.
    priority_1:
        Allocated priority 1 time.
    priority_2:
        Allocated priority 2 time.
    priority_3:
        Allocated priority 3 time.
    priority_4:
        Allocated priority 4 time.
    """

    partner_code: str
    partner_name: str
    tac_comment: Optional[str]
    priority_0: Quantity
    priority_1: Quantity
    priority_2: Quantity
    priority_3: Quantity
    priority_4: Quantity


class BlockVisit(BaseModel):
    """
    A block visit, i.e. an observation made for a block.

    Fields
    ------
    block_id:
       Database id of the observed block.
    block_name:
       Name of the observed block.
    observed_time:
       The nominal time spent on making the observation. This is the time charged for,
       and may differ from the actual time spent.
    priority:
       Priority of the block.
    max_lunar_phase:
       The maximum lunar phase allowed for the observation. The lunar phase is the
       illuminated fraction of the Moon, as a percentage. It is taken to be 0 if the
       Moon is below the horizon.
    target_names:
       Names of the observed targets.
    observation_night:
       Date of the observation night. This is the date when the night starts, i.e. all
       observation done during the same night have the same observation night date.
    status:
       Status of the observation, such "Accepted" or "Rejected".
    rejection_reason:
       Reason why the observation has been rejected.
    """

    block_id: int
    block_name: str
    observed_time: str
    priority: str
    max_lunar_phase: str
    target_names: List[str]
    observation_night: date
    status: str
    rejection_reason: Optional[str]


class Phase1Target(BaseModel):
    """
    Target details in a Phase 1 proposal.

    Fields
    ------
    name:
        Target name.
    right_ascension:
        Right ascension.
    declination:
        Declination.
    equinox:
        Equinox for the target coordinates.
    horizons_identifier:
        Identifier in the Horizons database for solar system targets.
    minimum_magnitude:
        Minimum magnitude of the target.
    maximum_magnitude:
        Maximum magnitude.
    target_type:
        Target type (broad category).
    target_subtype:
        Target subtype.
    is_optional:
        Whether the target is optional. i.e. whether it is part of a pool of targets
        from which only a subset needs to be observed.
    n_visits:
        The number of observations requested for the target.
    max_lunar_phase:
        The maximum lunar phase allowed for an observation of the target. The lunar
        phase is the illuminated fraction of the Moon, as a percentage. It is taken to
        be 0 if the Moon is below the horizon.
    ranking:
        Importance attributed by the Principal Investigator to observations of this
        target relative to other observations for the same proposal.
    night_count:
        The number of nights when the target can be observed, given the requested
        observation time and observation constraints.
    TODO: Comment on probabilities
    """

    name: str
    right_ascension: Optional[Angle]
    declination: Optional[Angle]
    equinox: float
    horizons_identifier: Optional[str]
    minimum_magnitude: float
    maximum_magnitude: float
    target_type: str
    target_subtype: str
    is_optional: bool
    n_visits: int
    max_lunar_phase: float
    ranking: str
    night_count: int
    moon_probability: int
    competition_probability: int
    observability_probability: int
    seeing_probability: int


class RequestedTime(BaseModel):
    total_requested_time: int
    minimum_useful_time: int
    time_comment: Optional[str]
    semester: Semester
    distribution: List


class Proposal(BaseModel):
    text_content: TextContent
    investigators: List[Investigator]
    block_visits: List[BlockVisit]
    observed_time: Dict[str, Any]
    time_allocations: TimeAllocations


class Phase1Proposal(BaseModel):
    text_content: TextContent
    investigators: List[Investigator]
    block_visits: List[BlockVisit]
    observed_time: Dict[str, Any]
    time_allocations: TimeAllocations