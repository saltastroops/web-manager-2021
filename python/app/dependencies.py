from datetime import datetime, timedelta, timezone
from functools import cache
from typing import Optional

from aiomysql import Pool
from fastapi import Depends, HTTPException
from jose import JWTError, jwt
from starlette import status

from app.models.pydantic import Semester, User
from app.service import user as user_service
from app.settings import Settings
from app.util.auth import ALGORITHM, OAuth2TokenOrCookiePasswordBearer
from app.util.database import DatabasePool


@cache  # for performance reasons, as the function is called for every request
def get_settings() -> Settings:
    """Get the Web Manager settings."""
    return Settings()


get_db = DatabasePool()


oauth2_scheme = OAuth2TokenOrCookiePasswordBearer("api/token")


async def get_current_user(
    settings: Settings = Depends(get_settings),
    token: str = Depends(oauth2_scheme),
    db: Pool = Depends(get_db),
) -> User:
    """Get the currently logged in user."""
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials.",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, settings.secret_key, algorithms=[ALGORITHM])
        username = payload.get("sub")
        if username is None:
            raise credentials_exception
    except JWTError:
        raise credentials_exception
    user = await user_service.get_user(username, db)
    if user is None:
        raise credentials_exception

    return User(**user.dict())  # turn UserInDB into User instance


def get_semester(
    year: Optional[int] = None, semester: Optional[int] = None
) -> Semester:
    if year is not None and semester is None:
        raise ValueError("A semester must be passed if a year is passed.")
    if semester is not None and year is None:
        raise ValueError("A year must be passed if a semester is passed.")

    if year is None and semester is None:
        return _current_semester()

    return Semester(year=year, semester=semester)


def _current_semester() -> Semester:
    """Get the current semester."""

    # Get the current time, taking into account that a semester starts and ends at noon
    # rather than midnight (UTC).
    t = datetime.now(tz=timezone.utc) - timedelta(hours=12)
    now_year = t.year
    now_month = t.month

    if now_month < 5:
        return Semester(year=now_year - 1, semester=2)
    elif now_month < 11:
        return Semester(year=now_year, semester=1)
    else:
        return Semester(year=now_year, semester=2)
