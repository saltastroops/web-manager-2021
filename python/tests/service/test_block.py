"""Tests for the block service."""
import pytest
from syrupy.assertion import SnapshotAssertion

from app.models.pydantic import Semester
from app.service import block as block_service


@pytest.mark.asyncio
async def test_get_block_raises_not_found_error_for_wrong_proposal_code() -> None:
    with pytest.raises(FileNotFoundError):
        await block_service.get_block(
            "non-existing",
            "abcd1234",
            Semester(year=2020, semester=1),
            "tests/__testdata__/proposals",
        )


@pytest.mark.asyncio
async def test_get_block_raises_not_found_error_for_wrong_block_code() -> None:
    with pytest.raises(FileNotFoundError):
        await block_service.get_block(
            "2020-2-SCI-065",
            "non-existing",
            Semester(year=2020, semester=1),
            "tests/__testdata__/proposals",
        )


@pytest.mark.asyncio
async def test_get_block_raises_not_found_error_for_wrong_semester() -> None:
    with pytest.raises(FileNotFoundError):
        await block_service.get_block(
            "2020-2-SCI-065",
            "abcd1234",
            Semester(year=2007, semester=1),
            "tests/__testdata__/proposals",
        )


@pytest.mark.asyncio
async def test_get_block_raises_not_found_error_for_wrong_base_directory() -> None:
    with pytest.raises(FileNotFoundError):
        await block_service.get_block(
            "2020-2-SCI-065",
            "abcd1234",
            Semester(year=2020, semester=1),
            "non-existing",
        )


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "proposal_code,block_code,semester",
    [
        ("2020-2-SCI-065", "abcd1234", Semester(year=2020, semester=1)),
        ("2020-2-SCI-065", "abcd1234", Semester(year=2020, semester=1)),
    ],
)
async def test_get_block_returns_correct_value(
    proposal_code: str, block_code: str, semester: Semester, snapshot: SnapshotAssertion
) -> None:
    """get_block returns the correct block content."""
    block = await block_service.get_block(
        proposal_code, block_code, semester, "tests/__testdata__/proposals"
    )
    assert block == snapshot
