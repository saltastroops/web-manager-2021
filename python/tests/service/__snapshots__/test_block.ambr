# name: test_get_block_returns_correct_value[2020-2-SCI-065-abcd1234-semester0]
  <class 'OrderedDict'> {
    '@id': 'Block-10',
    'BlockCode': 'fd6b10fe-6400-4676-ad6d-6283ac598295',
    'BlockSemester': <class 'list'> [
      <class 'OrderedDict'> {
        'Comments': 'P0 used to pick up a low-visibility needed phase.',
        'InPool': <class 'OrderedDict'> {
          'PoolCode': '9e840e44-3964-4575-aa3e-1c3ff5409375',
        },
        'Priority': '0',
        'Ranking': 'High',
        'Semester': '2',
        'Visits': '1',
        'Wait': <class 'OrderedDict'> {
          'Units': 'days',
          'Value': '1.0',
        },
        'Year': '2018',
      },
      <class 'OrderedDict'> {
        'Comments': 'P0 used to pick up a low-visibility needed phase.',
        'InPool': <class 'OrderedDict'> {
          'PoolCode': '9e840e44-3964-4575-aa3e-1c3ff5409375',
        },
        'Priority': '0',
        'Ranking': 'High',
        'Semester': '1',
        'Visits': '15',
        'Year': '2019',
      },
      <class 'OrderedDict'> {
        'Comments': '2400 secs + 600 sec acquisition. P0 used to pick up a low-visibility needed phase.',
        'InPool': <class 'OrderedDict'> {
          'PoolCode': '9e840e44-3964-4575-aa3e-1c3ff5409375',
        },
        'Priority': '0',
        'Ranking': 'High',
        'Semester': '2',
        'Visits': '20',
        'Year': '2019',
      },
      <class 'OrderedDict'> {
        'Comments': '2400 secs + 600 sec acquisition. P0 used to pick up a low-visibility needed phase.',
        'InPool': <class 'OrderedDict'> {
          'PoolCode': '9e840e44-3964-4575-aa3e-1c3ff5409375',
        },
        'Priority': '0',
        'Ranking': 'High',
        'Semester': '1',
        'Visits': '19',
        'Year': '2020',
      },
    ],
    'BlockType': 'Science',
    'Charged': 'true',
    'MaxVisits': '20',
    'MaximumLunarPhase': <class 'OrderedDict'> {
      'Units': '%',
      'Value': '100.0',
    },
    'MinimumLunarAngularDistance': <class 'OrderedDict'> {
      'Units': 'degrees',
      'Value': '30.0',
    },
    'Name': 'Some Target P0',
    'SeeingConstraint': <class 'OrderedDict'> {
      'Maximum': <class 'OrderedDict'> {
        'Units': 'arcseconds',
        'Value': '3.0',
      },
      'Minimum': <class 'OrderedDict'> {
        'Units': 'arcseconds',
        'Value': '0.1',
      },
    },
    'SubBlock': <class 'OrderedDict'> {
      'Iterations': '1',
      'Name': '[SubBlock]',
      'SubSubBlock': <class 'OrderedDict'> {
        'Iterations': '1',
        'Name': '[SubSubBlock]',
        'Pointing': <class 'OrderedDict'> {
          'Name': '[Pointing]',
          'Observation': <class 'OrderedDict'> {
            'Acquisition': <class 'OrderedDict'> {
              'Name': 'Some Target',
              'Target': <class 'OrderedDict'> {
                'Coordinates': <class 'OrderedDict'> {
                  'Declination': <class 'OrderedDict'> {
                    'Arcminutes': '57',
                    'Arcseconds': '38.7',
                    'Degrees': '23',
                    'Sign': '-',
                    'ns3': 'http://www.salt.ac.za/PIPT/Shared/1.4',
                  },
                  'Equinox': '2000.0',
                  'ProperMotionAndEpoch': <class 'OrderedDict'> {
                    'DeclinationDot': <class 'OrderedDict'> {
                      'Units': 'arcseconds/year',
                      'Value': '0.0055',
                    },
                    'Epoch': '2000-01-01T00:00:00.791Z',
                    'RightAscensionDot': <class 'OrderedDict'> {
                      'Units': 'arcseconds/year',
                      'Value': '-0.0024',
                    },
                    'ns3': 'http://www.salt.ac.za/PIPT/Shared/1.4',
                  },
                  'RightAscension': <class 'OrderedDict'> {
                    'Hours': '15',
                    'Minutes': '23',
                    'Seconds': '16.954',
                    'ns3': 'http://www.salt.ac.za/PIPT/Shared/1.4',
                  },
                },
                'FindingChart': <class 'OrderedDict'> {
                  'Path': 'Included/2e389f565d2cd4511cc3ab3191574929.pdf',
                },
                'MagnitudeRange': <class 'OrderedDict'> {
                  'Bandpass': 'V',
                  'Maximum': '14.71',
                  'Minimum': '14.71',
                },
                'Mandatory': 'true',
                'Name': 'Some Target',
                'PeriodicTargetEphemeris': <class 'OrderedDict'> {
                  'Pdot': <class 'OrderedDict'> {
                    'Units': 'days/day',
                    'Value': '0.0',
                  },
                  'Period': <class 'OrderedDict'> {
                    'Units': 'days',
                    'Value': '3.14159',
                  },
                  'TimeZero': <class 'OrderedDict'> {
                    'TimeBase': 'HJD',
                    'TimeValue': '2450540.14',
                  },
                },
                'TargetCode': '172529be-de04-42c7-96ed-e3d84f3d7cd2',
                'TargetType': 'WR*',
                'ns2': 'http://www.salt.ac.za/PIPT/Proposal/Shared/2.7',
              },
              'TelescopeConfig': <class 'OrderedDict'> {
                'Iterations': '1',
                'Name': 'Telescope Configuration',
                'PayloadConfig': <class 'OrderedDict'> {
                  'GuideMethod': 'None',
                  'Name': 'Payload Configuration',
                  'PellicleConfig': 'false',
                  'Salticam': <class 'OrderedDict'> {
                    'Cycles': '1',
                    'InCalibration': 'false',
                    'MinimumSN': '0',
                    'Name': 'Salticam Configuration',
                    'SalticamDetector': <class 'OrderedDict'> {
                      'DetMode': 'Normal',
                      'ExposureType': 'Science',
                      'Gain': 'Bright',
                      'Iterations': '1',
                      'PreBinCols': '2',
                      'PreBinRows': '2',
                      'ReadoutSpeed': 'Fast',
                    },
                    'SalticamProcedure': <class 'OrderedDict'> {
                      'SalticamFilterArray': <class 'OrderedDict'> {
                        'ExposureTime': <class 'OrderedDict'> {
                          'Units': 'seconds',
                          'Value': '1.0',
                        },
                        'Filter': 'Johnson V',
                      },
                    },
                    'ns4': 'http://www.salt.ac.za/PIPT/Salticam/Phase2/1.9',
                  },
                  'Type': 'Acquisition',
                  'WaitOnOtherInstrument': 'true',
                },
              },
            },
            'Name': 'Some Target',
            'PipelineConfig': <class 'OrderedDict'> {
              'Notification': 'Normal',
              'ReductionDocumentationFormat': 'HTML',
            },
            'TelescopeConfig': <class 'OrderedDict'> {
              'Iterations': '1',
              'Name': '[Telescope Configuration]',
              'OnSkyPositionAngle': <class 'OrderedDict'> {
                'Units': 'degrees',
                'Value': '0.0',
              },
              'PayloadConfig': <class 'OrderedDict'> {
                'GuideMethod': 'Default',
                'Name': '[Payload Configuration]',
                'PellicleConfig': 'false',
                'Rss': <class 'OrderedDict'> {
                  'Cycles': '3',
                  'InCalibration': 'false',
                  'MinimumSN': '0',
                  'Name': 'fainter G0900 spectropolarimetry',
                  'RssCalibration': <class 'OrderedDict'> {
                    'RssArc': <class 'OrderedDict'> {
                      'ArcExposureTime': <class 'OrderedDict'> {
                        'Units': 'seconds',
                        'Value': '5.0',
                      },
                      'ArcRequirement': 'After science',
                      'DoneBetweenDitheringSteps': 'false',
                      'Iterations': '1',
                      'UseDefaultArcLamp': None,
                    },
                  },
                  'RssConfig': <class 'OrderedDict'> {
                    'FilterId': 'pc03850',
                    'FocusPosition': <class 'OrderedDict'> {
                      'Units': 'microns',
                      'Value': '1000',
                    },
                    'Mode': <class 'OrderedDict'> {
                      'Spectroscopy': <class 'OrderedDict'> {
                        'ArtStation': '39_30.25',
                        'Grating': 'pg0900',
                        'GratingAngle': <class 'OrderedDict'> {
                          'Units': 'degrees',
                          'Value': '15.125',
                        },
                      },
                    },
                    'Polarimetry': <class 'OrderedDict'> {
                      'BeamsplitterOrientation': 'normal',
                    },
                    'SlitMask': <class 'OrderedDict'> {
                      'PredefinedMask': <class 'OrderedDict'> {
                        'Barcode': 'PL0125P001',
                      },
                    },
                  },
                  'RssDetector': <class 'OrderedDict'> {
                    'DetMode': 'Normal',
                    'DetectorCalculation': 'None',
                    'ExposureTime': <class 'OrderedDict'> {
                      'Units': 'seconds',
                      'Value': '157.0',
                    },
                    'ExposureType': 'Science',
                    'Gain': 'Bright',
                    'Iterations': '1',
                    'PostShuffle': '0',
                    'PreBinCols': '2',
                    'PreBinRows': '4',
                    'PreShuffle': '0',
                    'Prepare': 'true',
                    'ReadoutSpeed': 'Fast',
                  },
                  'RssNonChargedCalibration': <class 'OrderedDict'> {
                    'RssSpectroscopicFlat': <class 'OrderedDict'> {
                      'Iterations': '5',
                      'Type': 'Lamp',
                    },
                  },
                  'RssProcedure': <class 'OrderedDict'> {
                    'WaveplatePattern': <class 'OrderedDict'> {
                      'PatternStep': <class 'list'> [
                        <class 'OrderedDict'> {
                          'HWStation': '0_0',
                        },
                        <class 'OrderedDict'> {
                          'HWStation': '4_45.00',
                        },
                        <class 'OrderedDict'> {
                          'HWStation': '2_22.50',
                        },
                        <class 'OrderedDict'> {
                          'HWStation': '6_67.50',
                        },
                      ],
                    },
                  },
                  'ns5': 'http://www.salt.ac.za/PIPT/RSS/Phase2/2.3',
                },
                'Type': 'Science',
                'WaitOnOtherInstrument': 'true',
              },
            },
          },
        },
      },
    },
    'Transparency': 'Thin cloud',
    'ns1': 'http://www.salt.ac.za/PIPT/Proposal/Phase2/4.8',
  }
---
# name: test_get_block_returns_correct_value[2020-2-SCI-065-abcd1234-semester1]
  <class 'OrderedDict'> {
    '@id': 'Block-10',
    'BlockCode': 'fd6b10fe-6400-4676-ad6d-6283ac598295',
    'BlockSemester': <class 'list'> [
      <class 'OrderedDict'> {
        'Comments': 'P0 used to pick up a low-visibility needed phase.',
        'InPool': <class 'OrderedDict'> {
          'PoolCode': '9e840e44-3964-4575-aa3e-1c3ff5409375',
        },
        'Priority': '0',
        'Ranking': 'High',
        'Semester': '2',
        'Visits': '1',
        'Wait': <class 'OrderedDict'> {
          'Units': 'days',
          'Value': '1.0',
        },
        'Year': '2018',
      },
      <class 'OrderedDict'> {
        'Comments': 'P0 used to pick up a low-visibility needed phase.',
        'InPool': <class 'OrderedDict'> {
          'PoolCode': '9e840e44-3964-4575-aa3e-1c3ff5409375',
        },
        'Priority': '0',
        'Ranking': 'High',
        'Semester': '1',
        'Visits': '15',
        'Year': '2019',
      },
      <class 'OrderedDict'> {
        'Comments': '2400 secs + 600 sec acquisition. P0 used to pick up a low-visibility needed phase.',
        'InPool': <class 'OrderedDict'> {
          'PoolCode': '9e840e44-3964-4575-aa3e-1c3ff5409375',
        },
        'Priority': '0',
        'Ranking': 'High',
        'Semester': '2',
        'Visits': '20',
        'Year': '2019',
      },
      <class 'OrderedDict'> {
        'Comments': '2400 secs + 600 sec acquisition. P0 used to pick up a low-visibility needed phase.',
        'InPool': <class 'OrderedDict'> {
          'PoolCode': '9e840e44-3964-4575-aa3e-1c3ff5409375',
        },
        'Priority': '0',
        'Ranking': 'High',
        'Semester': '1',
        'Visits': '19',
        'Year': '2020',
      },
    ],
    'BlockType': 'Science',
    'Charged': 'true',
    'MaxVisits': '20',
    'MaximumLunarPhase': <class 'OrderedDict'> {
      'Units': '%',
      'Value': '100.0',
    },
    'MinimumLunarAngularDistance': <class 'OrderedDict'> {
      'Units': 'degrees',
      'Value': '30.0',
    },
    'Name': 'Some Target P0',
    'SeeingConstraint': <class 'OrderedDict'> {
      'Maximum': <class 'OrderedDict'> {
        'Units': 'arcseconds',
        'Value': '3.0',
      },
      'Minimum': <class 'OrderedDict'> {
        'Units': 'arcseconds',
        'Value': '0.1',
      },
    },
    'SubBlock': <class 'OrderedDict'> {
      'Iterations': '1',
      'Name': '[SubBlock]',
      'SubSubBlock': <class 'OrderedDict'> {
        'Iterations': '1',
        'Name': '[SubSubBlock]',
        'Pointing': <class 'OrderedDict'> {
          'Name': '[Pointing]',
          'Observation': <class 'OrderedDict'> {
            'Acquisition': <class 'OrderedDict'> {
              'Name': 'Some Target',
              'Target': <class 'OrderedDict'> {
                'Coordinates': <class 'OrderedDict'> {
                  'Declination': <class 'OrderedDict'> {
                    'Arcminutes': '57',
                    'Arcseconds': '38.7',
                    'Degrees': '23',
                    'Sign': '-',
                    'ns3': 'http://www.salt.ac.za/PIPT/Shared/1.4',
                  },
                  'Equinox': '2000.0',
                  'ProperMotionAndEpoch': <class 'OrderedDict'> {
                    'DeclinationDot': <class 'OrderedDict'> {
                      'Units': 'arcseconds/year',
                      'Value': '0.0055',
                    },
                    'Epoch': '2000-01-01T00:00:00.791Z',
                    'RightAscensionDot': <class 'OrderedDict'> {
                      'Units': 'arcseconds/year',
                      'Value': '-0.0024',
                    },
                    'ns3': 'http://www.salt.ac.za/PIPT/Shared/1.4',
                  },
                  'RightAscension': <class 'OrderedDict'> {
                    'Hours': '15',
                    'Minutes': '23',
                    'Seconds': '16.954',
                    'ns3': 'http://www.salt.ac.za/PIPT/Shared/1.4',
                  },
                },
                'FindingChart': <class 'OrderedDict'> {
                  'Path': 'Included/2e389f565d2cd4511cc3ab3191574929.pdf',
                },
                'MagnitudeRange': <class 'OrderedDict'> {
                  'Bandpass': 'V',
                  'Maximum': '14.71',
                  'Minimum': '14.71',
                },
                'Mandatory': 'true',
                'Name': 'Some Target',
                'PeriodicTargetEphemeris': <class 'OrderedDict'> {
                  'Pdot': <class 'OrderedDict'> {
                    'Units': 'days/day',
                    'Value': '0.0',
                  },
                  'Period': <class 'OrderedDict'> {
                    'Units': 'days',
                    'Value': '3.14159',
                  },
                  'TimeZero': <class 'OrderedDict'> {
                    'TimeBase': 'HJD',
                    'TimeValue': '2450540.14',
                  },
                },
                'TargetCode': '172529be-de04-42c7-96ed-e3d84f3d7cd2',
                'TargetType': 'WR*',
                'ns2': 'http://www.salt.ac.za/PIPT/Proposal/Shared/2.7',
              },
              'TelescopeConfig': <class 'OrderedDict'> {
                'Iterations': '1',
                'Name': 'Telescope Configuration',
                'PayloadConfig': <class 'OrderedDict'> {
                  'GuideMethod': 'None',
                  'Name': 'Payload Configuration',
                  'PellicleConfig': 'false',
                  'Salticam': <class 'OrderedDict'> {
                    'Cycles': '1',
                    'InCalibration': 'false',
                    'MinimumSN': '0',
                    'Name': 'Salticam Configuration',
                    'SalticamDetector': <class 'OrderedDict'> {
                      'DetMode': 'Normal',
                      'ExposureType': 'Science',
                      'Gain': 'Bright',
                      'Iterations': '1',
                      'PreBinCols': '2',
                      'PreBinRows': '2',
                      'ReadoutSpeed': 'Fast',
                    },
                    'SalticamProcedure': <class 'OrderedDict'> {
                      'SalticamFilterArray': <class 'OrderedDict'> {
                        'ExposureTime': <class 'OrderedDict'> {
                          'Units': 'seconds',
                          'Value': '1.0',
                        },
                        'Filter': 'Johnson V',
                      },
                    },
                    'ns4': 'http://www.salt.ac.za/PIPT/Salticam/Phase2/1.9',
                  },
                  'Type': 'Acquisition',
                  'WaitOnOtherInstrument': 'true',
                },
              },
            },
            'Name': 'Some Target',
            'PipelineConfig': <class 'OrderedDict'> {
              'Notification': 'Normal',
              'ReductionDocumentationFormat': 'HTML',
            },
            'TelescopeConfig': <class 'OrderedDict'> {
              'Iterations': '1',
              'Name': '[Telescope Configuration]',
              'OnSkyPositionAngle': <class 'OrderedDict'> {
                'Units': 'degrees',
                'Value': '0.0',
              },
              'PayloadConfig': <class 'OrderedDict'> {
                'GuideMethod': 'Default',
                'Name': '[Payload Configuration]',
                'PellicleConfig': 'false',
                'Rss': <class 'OrderedDict'> {
                  'Cycles': '3',
                  'InCalibration': 'false',
                  'MinimumSN': '0',
                  'Name': 'fainter G0900 spectropolarimetry',
                  'RssCalibration': <class 'OrderedDict'> {
                    'RssArc': <class 'OrderedDict'> {
                      'ArcExposureTime': <class 'OrderedDict'> {
                        'Units': 'seconds',
                        'Value': '5.0',
                      },
                      'ArcRequirement': 'After science',
                      'DoneBetweenDitheringSteps': 'false',
                      'Iterations': '1',
                      'UseDefaultArcLamp': None,
                    },
                  },
                  'RssConfig': <class 'OrderedDict'> {
                    'FilterId': 'pc03850',
                    'FocusPosition': <class 'OrderedDict'> {
                      'Units': 'microns',
                      'Value': '1000',
                    },
                    'Mode': <class 'OrderedDict'> {
                      'Spectroscopy': <class 'OrderedDict'> {
                        'ArtStation': '39_30.25',
                        'Grating': 'pg0900',
                        'GratingAngle': <class 'OrderedDict'> {
                          'Units': 'degrees',
                          'Value': '15.125',
                        },
                      },
                    },
                    'Polarimetry': <class 'OrderedDict'> {
                      'BeamsplitterOrientation': 'normal',
                    },
                    'SlitMask': <class 'OrderedDict'> {
                      'PredefinedMask': <class 'OrderedDict'> {
                        'Barcode': 'PL0125P001',
                      },
                    },
                  },
                  'RssDetector': <class 'OrderedDict'> {
                    'DetMode': 'Normal',
                    'DetectorCalculation': 'None',
                    'ExposureTime': <class 'OrderedDict'> {
                      'Units': 'seconds',
                      'Value': '157.0',
                    },
                    'ExposureType': 'Science',
                    'Gain': 'Bright',
                    'Iterations': '1',
                    'PostShuffle': '0',
                    'PreBinCols': '2',
                    'PreBinRows': '4',
                    'PreShuffle': '0',
                    'Prepare': 'true',
                    'ReadoutSpeed': 'Fast',
                  },
                  'RssNonChargedCalibration': <class 'OrderedDict'> {
                    'RssSpectroscopicFlat': <class 'OrderedDict'> {
                      'Iterations': '5',
                      'Type': 'Lamp',
                    },
                  },
                  'RssProcedure': <class 'OrderedDict'> {
                    'WaveplatePattern': <class 'OrderedDict'> {
                      'PatternStep': <class 'list'> [
                        <class 'OrderedDict'> {
                          'HWStation': '0_0',
                        },
                        <class 'OrderedDict'> {
                          'HWStation': '4_45.00',
                        },
                        <class 'OrderedDict'> {
                          'HWStation': '2_22.50',
                        },
                        <class 'OrderedDict'> {
                          'HWStation': '6_67.50',
                        },
                      ],
                    },
                  },
                  'ns5': 'http://www.salt.ac.za/PIPT/RSS/Phase2/2.3',
                },
                'Type': 'Science',
                'WaitOnOtherInstrument': 'true',
              },
            },
          },
        },
      },
    },
    'Transparency': 'Thin cloud',
    'ns1': 'http://www.salt.ac.za/PIPT/Proposal/Phase2/4.8',
  }
---
